### ISSUE TITLE
Does nominal/categorical data need to be inherently numeric?

### ISSUE THREAD

**Issue by pynkfloydfan** [KacperKubara/heterogeneous_dist_metrics#5]
Hi Kacper

This looks to be a great package potentially but I encountered a problem when trying to pass a dataframe with categorical information in string format.  The traceback gave the error:

"unsupported operand type(s) for -: 'str' and 'str'

In your example you only use a numeric dataset even though you declare 2 columns as categorical.  Should I presume then that we need to apply a LabelEncoder to any data that is not inherently numeric before using these functions?

Thanks
Lloyd

**Comment by KacperKubara** [comment]
Hey Lloyd,
thanks for the message. If you have strings, they should be label encoded or onehotencoded before using the metrics. Actually, it might be worth throwing some error and adding a note to docs about it- I will have look at it when I will have a bit more spare time!

**Comment by pynkfloydfan** [comment]
Thanks for confirming Kacper!

I presume that OneHotEncoded and LabelEncoded data would not produce the same distance result: if two items have the same value then their distance would be 0 but if they had different values it would be 1 if LabelEncoded and 2 if OneHotEncoded wouldn't it?

Intuitively it feels like LabelEncoding would be the most appropriate way to go?

Maybe there could be an option to LabelEncode the categorical columns as part of the build procedure?

What do you think?

**Comment by KacperKubara** [comment]
Hello Lloyd,
I think that you're right, label encoder should be the way to go as the one hot encoder would "double" the distance because we will be effectively creating another column. 

I am not sure if it is worth including the label encoding as a part of the build structure because it probably means that I will need to import Scikit-Learn preprocessing module. The idea of the package is rather to have only pure Numpy implementation of the algorithm (to make it fast and compact).

Having said so, I think including some preprocessing example in the library would be helpful. What do you think?

**Comment by pynkfloydfan** [comment]
Hi Kacper

I hope you don’t mind me mailing you directly, I didn’t want to put this on the same thread in github.

Regarding speed, I've been testing the heom metric as a custom distance input to both the HDBSCAN clustering algorithm in the hdbscan package and the Scikit-Learn implementation of t-SNE dimensionality reduction.  Whilst they work, the increase in time taken to calculate has been quite dramatic, to the extent that I can really only test a fraction of my data.  For example, whilst it took about 30mins to run the hdbscan algorithm on a 36k row dataset using Euclidean distance (and where I’d one-hot encoded 2 categorical columns), it was still running 8 hours later using heom.

Unfortunately also, the soft clustering and predictive clustering capabilities of hdbscan don't work with a custom distance matrix either.

I think it’s fantastic what you’ve done: you've solved a significant problem by providing a robust distance calculation for categorical/mixed data that just doesn't exist in python, but to get the best benefit from them I'm wondering if they need to be implemented in the main sklearn package?

Thanks for all your hard work
Lloyd

**Comment by pynkfloydfan** [comment]
Hi Kacper, yes I think that would be really helpful.  While you're at it, an example of how to implement the HVDM metric would be useful too, I didn't quite understand what is expected for 'y' :-)

**Comment by KacperKubara** [comment]
Hey Lloyd,
Ideally, this algorithm should be implemented in the Scikit-Learn package, because the overhead of calling HEOM from Distython is ridiculously big. In Scikit-Learn, they also use Numba for the algorithm implementations which pre-compiles the code to C to make it even faster than the Numpy implementation. It might be actually worth creating a feature request in the Scikit-Learn repo to ask for the implementation of the heterogeneous distance metrics? This package might be a good starting point for that. 

Regarding the modifications to the Distython, I will aim to do some changes next week as I will have a bit more spare time!

**Comment by pynkfloydfan** [comment]
Dear Kacper

Apologies for the late reply: you really are a super kind fellow and thanks for engaging with me.

I’ve submitted a feature request today (#15894) so fingers crossed and let’s see what the authors think.

All the best
Lloyd

**Comment by KacperKubara** [comment]
Hey Lloyd!
I have changed the package so now it should throw TypeError if you provide data which include columns with strings: https://github.com/KacperKubara/distython/blob/f04fff9231eb585d647ad21e35683724fb5c89bc/HEOM.py#L32-L35 

When I was reproducing your issue, I realized that you probably used pandas DataFrame instead of Numpy array, am I right? The Pandas DataFrame should be converted to Numpy before using Distython as it might produce unexpected behaviour when using the metrics. From what I remember, you can do it with `X.to_numpy()` 

The changes are still in dev branch so the package from PyPI is not yet updated. I will aim to test it first and then update the package, together with adding some additional examples on how to use the package, after holidays :) 

Have a merry christmas!
Kacper
